uanets/models/interface.py

```
@@ -0,0 +1,92 @@
+        need to sample to compute them, in these cases ``num_samples`` and ``seed`` can be used.
+
+        Models implementing this function should decorate the method with
+        :meth:`~check_shapes.inherit_check_shapes`, so that check_shapes specified above are
+        inherited.
+
+        :param x:
+            Input locations at which to compute mean and variance.
+        :param num_samples:
+            Some models might need to sample to compute mean and the variance. In these
+            cases this argument can be used to set the number of samples based on which mean
+            and variance will be computed.
+        :param seed:
+            Some models might need to sample to compute mean and the variance. In these cases
+            seed can be fixed to produce deterministic results.
+        :return: The mean and variance of the independent marginal distributions at each point in
+            ``x``.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    @check_shapes(
+        "x: [batch..., N, D]",
+        "return: [batch..., N, P] if (num_samples is None)",
+        "return: [batch..., S, N, P] if (num_samples is not None)",
+    )
+    def sample(
+        self,
+        x: tf.Tensor,
+        num_samples: Optional[int] = None,
+        seed: Optional[int] = None,
+    ) -> tf.Tensor:
+        """
+        Return ``num_samples`` samples from the independent marginal distributions at ``x``.
+
+        Models implementing this function should decorate the method with
+        :meth:`~check_shapes.inherit_check_shapes`, so that check_shapes specified above are
+        inherited.
+
+        :param x:
+            Input locations at which to draw samples.
+        :param num_samples:
+            Number of samples to draw.
+            If `None`, a single sample is drawn and the return shape is [..., N, P],
+            for any positive integer the return shape contains an extra batch
```

I have a feeling this is similar to what is done in `gpflow`?, but I wonder if it's simpler to just default to a sample size of 1 and return something with [..., 1, N, P] shape?  I guess I'm curious as to why `gpflow` made this decision...makes the output easier to use?

---

yes, thats just copied from gpflow, it also striked me as unnecessary complication and wondered why they went with it...

---

I'll switch to `[..,1,N,P]`, thats more sensible